fw/sock_clnt.c

```diff
@@ -157,264 +160,45 @@ tfw_cli_conn_send(TfwCliConn *cli_conn, TfwMsg *msg)
-		}
-	}
-
-	if (FRAME_DATA_SHOULD_BE_MADE(flags)) {
-		if (stream->rem_wnd <= 0 || h2->rem_wnd <= 0
-		    || *limit - stream->xmit.processed <= FRAME_HEADER_SIZE) {
-			if (headers_was_done)
-				goto update_limit;
-			r = -ENOMEM;
-			goto ret;
-		}
-
-		r = tfw_h2_make_data_frames(sk, skb, h2, stream, mss_now,
-					    *limit - stream->xmit.processed,
-					    &tmp_truesize);
-		if (unlikely(r)) {
-			T_WARN("%s: failed to make data frames (%d)",
-			       __func__, r);
-			if (r == -ENOMEM && headers_was_done) {
-				r = 0;
-				goto update_limit;
-			}
-			goto ret;
-		}
-
-		truesize += tmp_truesize;
-		tmp_truesize = 0;
-
-		/*
-		 * We clear this flag to prevent it's copying
-		 * during skb splitting.
-		 */
-		if (!stream->xmit.b_len) {
-			skb_clear_tfw_flag(skb, SS_F_HTTT2_FRAME_DATA);
-			TFW_H2_STREAM_SEND_PROCESS(h2, stream, HTTP2_DATA);
-		}
-	}
-
-update_limit:
-	if (FRAME_HEADERS_OR_DATA_SHOULD_BE_MADE(flags)
-	    && stream && stream->xmit.nskbs == 1)
-		*limit = stream->xmit.processed;
-
-	if (skb->len > *limit) {
-		unsigned short saved_flags = skb_tfw_flags(skb);
-
-		/*
-		 * Hacky way to clear flags of skb that will be created after
-		 * splitting such skb must be with cleared flags, but
-		 * current skb must be with already set flags.
-		 */
-		skb->tfw_cb.flags &= (unsigned short)(~TEMPESTA_SKB_FLAG_CLEAR_MASK);
-		r = tso_fragment(sk, skb, *limit, mss_now,
-				 sk_gfp_mask(sk, GFP_ATOMIC));
-		skb->tfw_cb.flags = saved_flags;
-	}
-
-ret:
-	/* Reinit stream xmit context. */
-	if (stream)
-		*nskbs = !r ? stream->xmit.nskbs : 0;
-
-	/*
-	 * Since we add some data to skb, we should adjust the socket write
-	 * memory both in case of success and in case of failure.
-	 */
-	if (unlikely(ss_add_overhead(sk, truesize))) {
-		T_WARN("%s: failed to add overhead to current TCP "
-		       "socket control data.", __func__);
-		/*
-		 * In case of previous error return it,
-		 * otherwise return -ENOMEM.
-		 */
-		r = r ? r : -ENOMEM;
-	}
+	assert_spin_locked(&sk->sk_lock.slock);
 
-	if (unlikely(r) && r != -ENOMEM) {
-		if (stream)
-			tfw_h2_stream_add_closed(h2, stream);
+	r = tfw_tls_encrypt(sk, skb, mss_now, limit);
+	if (unlikely(r) && r != -ENOMEM)
 		/*
-		 * We can not send unencrypted data and can not normally close
-		 * the socket with FIN since we're in progress on sending from
-		 * the write queue.
+		 * We can not send unencrypted data and can not normally close the
+		 * socket with FIN since we're in progress on sending from the write
+		 * queue.
 		 */
 		ss_close(sk, SS_F_ABORT);
-	}
-
-	if (likely(!r))
-		skb_set_tfw_flags(skb, SS_F_HTTP2_FRAME_PREPARED);
-
 	return r;
+}
 
-#undef TFW_H2_STREAM_SEND_PROCESS
-#undef CHECK_STREAM_IS_PRESENT
-#undef FRAME_ALREADY_PREPARED
-#undef FRAME_HEADERS_OR_DATA_SHOULD_BE_MADE
-#undef FRAME_DATA_SHOULD_BE_MADE
-#undef FRAME_HEADERS_SHOULD_BE_MADE
+static inline unsigned long
+tfw_sk_calc_cwnd_awail(struct sock *sk, unsigned int mss_now)
```

Comment:
I must draw your attention to this function. You are calculating `snd_cwnd`, not `snd_wnd`. If you want to place into the queue the exact size that the client can receive, you should take into account `snd_wnd`. If there are no losses in the network, each time you try to fill the write_queue, `snd_cwnd` segments will be queued, and it can be much bigger than `snd_wnd`